libs/sdk-backend-spi/src/workspace/dashboards/layout/builder/layout.ts

```diff
@@ -110,48 +208,67 @@ export class FluidLayoutBuilder<TContent> implements IFluidLayoutBuilder<TConten
             this.removeRow(rowsToRemove.index());
         }
         return this;
-    };
+    }
 
-    public modifyRows = (
-        modify: FluidLayoutRowModifications<TContent>,
-        selector: FluidLayoutRowsSelector<TContent> = (rows) => rows.all(),
-    ): this => {
-        const rowsToModify = selector(this.facade().rows());
+    public removeEmptyRows(): this {
+        return this.removeRows((rows) => rows.filter((row) => row.isEmpty()));
+    }
+
+    public modifyRows(
+        modify: FluidLayoutRowModifications<
+            TContent,
+            TColumn,
+            TRow,
+            TRowFacade,
+            TColumnFacade,
+            TColumnsFacade,
+            TColumnBuilder,
+            TRowBuilder
+        >,
+        selector: FluidLayoutRowsSelector<TContent, TRow, TRowFacade, TRowsFacade> = (rows) => rows.all(),
+    ): this {
+        const rowsToModify = selector(this.facade().rows() as TRowsFacade);
         if (isArray(rowsToModify)) {
-            this.setLayout((layout) => {
-                const updatedRows = [...layout.rows];
-                rowsToModify.forEach((row) => {
-                    const rowBuilder = FluidLayoutRowBuilder.for(this, row.index());
-                    const modifiedRow = modify(rowBuilder, row).build();
-                    updatedRows[row.index()] = modifiedRow;
-                });
-                return {
-                    ...layout,
-                    rows: updatedRows,
-                };
+            rowsToModify.forEach((row) => {
+                this.modifyRow(row.index(), modify);
             });
         } else if (rowsToModify) {
             this.modifyRow(rowsToModify.index(), modify);
         }
         return this;
-    };
+    }
 
-    public setLayout(valueOrUpdateCallback: ValueOrUpdateCallback<IFluidLayout<TContent>>): this {
+    public setLayout = (valueOrUpdateCallback: ValueOrUpdateCallback<TLayout>): this => {
         const updatedLayout = resolveValueOrUpdateCallback(valueOrUpdateCallback, this.build());
-        this.layoutFacade = FluidLayoutFacade.for(updatedLayout);
+        this.layoutFacade = this.layoutFacadeConstructor(updatedLayout);
         return this;
-    }
+    };
 
-    public facade(): IFluidLayoutFacade<TContent> {
+    public facade(): TLayoutFacade {
         return this.layoutFacade;
     }
 
-    public modify = (modifications: FluidLayoutModifications<TContent>): this => {
+    public modify(
+        modifications: FluidLayoutModifications<
+            TContent,
+            TLayout,
+            TRow,
+            TColumn,
+            TRowFacade,
```

Comment:
This is a bit strange, I would say that the Facade types should be inlined and derived from "their" content types. Same for builders.

Reply:
So ideally (imo) it would be just 

```ts
FluidLayoutModifications<
             TContent,
             TLayout,
             TRow,
             TColumn,
             this
>
```

Reply:
Would break inference, not worth it